File: src/gui/render/c/d2d.cpp
#include "d2d.h"
#include <d2d1.h>
#include <dwrite.h>
#include <stdio.h>
#include <stdbool.h>

struct D2DContext {
    HWND window;
    ID2D1Factory *factory;
    ID2D1HwndRenderTarget *rt;
    IDWriteFactory *dwrite_factory;
    // TODO: Support loading an arbitrary number of these
    IDWriteTextFormat *text_format;

    ID2D1SolidColorBrush *solid_color;

    void *user;
};

static int globalOpenGUICount = 0;

static void setBrushColor(D2DContext *ctx, D2DColor color) {
    if (!ctx->solid_color) {
        HRESULT hr = ctx->rt->CreateSolidColorBrush(
            {.r = color.r, .g = color.g, .b = color.b, .a = color.a},
            &ctx->solid_color
        );
        if (!SUCCEEDED(hr)) {
            log_err_return("Brush create failed");
        }
    } else {
        ctx->solid_color->SetColor({.r = color.r, .g = color.g, .b = color.b, .a = color.a});
    }
}

void d2dl_beginDrawing(D2DContext *ctx) {
    ctx->rt->BeginDraw();
}

void d2dl_endDrawing(D2DContext *ctx) {
    HRESULT hr = ctx->rt->EndDraw();
    if (!SUCCEEDED(hr)) {
        log_err_return("EndDraw failed for some reason");
    }
}

void d2dl_clear(D2DContext *ctx, D2DColor color) {
    ctx->rt->Clear({.r = color.r, .g = color.g, .b = color.b, .a = color.a});
}

void d2dl_drawRect(D2DContext *ctx, D2DRect rect, float stroke_width, D2DColor color) {
    setBrushColor(ctx, color);
    ctx->rt->DrawRectangle(
        {.left = rect.left, .top = rect.top, .right = rect.right, .bottom = rect.bottom},
        ctx->solid_color,
        stroke_width
    );
}

void d2dl_fillRect(D2DContext *ctx, D2DRect rect, D2DColor color) {
    setBrushColor(ctx, color);
    ctx->rt->FillRectangle(
        {.left = rect.left, .top = rect.top, .right = rect.right, .bottom = rect.bottom},
        ctx->solid_color
    );
}

void d2dl_drawRoundedRect(D2DContext *ctx, D2DRect rect, float radius,
                          float stroke_width, D2DColor color) {
    setBrushColor(ctx, color);
    D2D1_RECT_F reg_rect = {.left = rect.left, .top = rect.top, .right = rect.right, .bottom = rect.bottom};
    ctx->rt->DrawRoundedRectangle(
        {.rect = reg_rect, .radiusX = radius, .radiusY = radius},
        ctx->solid_color, stroke_width
    );
}

void d2dl_fillRoundedRect(D2DContext *ctx, D2DRect rect, float radius,
                          D2DColor color) {
    setBrushColor(ctx, color);
    D2D1_RECT_F reg_rect = {.left = rect.left, .top = rect.top, .right = rect.right, .bottom = rect.bottom};
    ctx->rt->FillRoundedRectangle(
        {.rect = reg_rect, .radiusX = radius, .radiusY = radius},
        ctx->solid_color
    );
}

void d2dl_drawEllipse(D2DContext *ctx, D2DPoint center, float rx, float ry,
                      float stroke_width, D2DColor color) {
    setBrushColor(ctx, color);
    ctx->rt->DrawEllipse(
        {{center.x, center.y}, rx, ry},
        ctx->solid_color,
        stroke_width
    );
}

void d2dl_fillEllipse(D2DContext *ctx, D2DPoint center, float rx, float ry,
                      D2DColor color) {
    setBrushColor(ctx, color);
    ctx->rt->FillEllipse(
        {{center.x, center.y}, rx, ry},
        ctx->solid_color
    );
}

void d2dl_loadFont(D2DContext *ctx, const wchar_t *font_name, float size) {
    // TODO: Implement loading different font styles
    ctx->dwrite_factory->CreateTextFormat(
        font_name, NULL, DWRITE_FONT_WEIGHT_NORMAL,
        DWRITE_FONT_STYLE_NORMAL, DWRITE_FONT_STRETCH_NORMAL, size, L"",
        &ctx->text_format);
}

void d2dl_drawText(D2DContext *ctx, const wchar_t *text, D2DRect rect, D2DColor color) {
    setBrushColor(ctx, color);
    ctx->text_format->SetTextAlignment(DWRITE_TEXT_ALIGNMENT_CENTER);
    ctx->text_format->SetParagraphAlignment(DWRITE_PARAGRAPH_ALIGNMENT_CENTER);
    size_t len = wcslen(text);
    ctx->rt->DrawText(text, len, ctx->text_format,
        {.left = rect.left, .top = rect.top, .right = rect.right, .bottom = rect.bottom},
        ctx->solid_color
    );
}

D2DSize d2dl_getRenderSize(D2DContext *ctx) {
    auto size = ctx->rt->GetSize();
    return {.width = size.width, .height = size.height};
}

void d2dl_setRenderSize(D2DContext *ctx, D2DSize size) {
    HRESULT hr = ctx->rt->Resize({.width = (u32)size.width, .height = (u32)size.height});
    if (!SUCCEEDED(hr)) {
        log_err_return("Resize failed");
    }
}

static void d2dl_createFactory(D2DContext *ctx) {
  HRESULT hr = D2D1CreateFactory(D2D1_FACTORY_TYPE_SINGLE_THREADED,
                                 &ctx->factory);
  if (SUCCEEDED(hr)) {
    return;
  } else {
    log_err_return("ID2D1Factory creation failed");
  }
}

static void d2dl_createDirectWriteFactory(D2DContext *ctx) {
    HRESULT hr = DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, 
                                     __uuidof(ctx->dwrite_factory),
                                     (IUnknown**)&ctx->dwrite_factory);
    if (!SUCCEEDED(hr)) {
        log_err_return("DWriteFactory creation failed")
    }
}


static void d2dl_createRenderTarget(D2DContext *ctx) {
    RECT rc;
    GetClientRect(ctx->window, &rc);

    ID2D1HwndRenderTarget *rt = NULL;

    HRESULT hr = ctx->factory->CreateHwndRenderTarget(
      D2D1::RenderTargetProperties(),
      D2D1::HwndRenderTargetProperties(
          ctx->window, D2D1::SizeU(rc.right - rc.left, rc.bottom - rc.top)), &rt);

    if (SUCCEEDED(hr)) {
        ctx->rt = rt;
    }
    else {
        log_err_return("CreateHwndRenderTarget failed")
    }
}

D2DContext *d2dl_initDesktopWindow(int width, int height, const char *window_title) {
    D2DContext *ctx = (D2DContext*)calloc(0, sizeof(D2DContext));
    if (ctx == NULL) {
        log_err_return_null("D2DContext allocation failed")
    }
    memset(ctx, 0, sizeof(D2DContext));
    d2dl_createFactory(ctx);
    d2dl_createDirectWriteFactory(ctx);
    HWND hwnd = CreateWindow(window_title, window_title, WS_OVERLAPPEDWINDOW,
                             CW_USEDEFAULT, CW_USEDEFAULT, width,
                         height, NULL, NULL, NULL, ctx);
    if (hwnd == NULL) {
        log_err_return_null("CreateWindow failed");
    }
    ctx->window = hwnd;
    d2dl_createRenderTarget(ctx);

    ShowWindow(hwnd, SW_SHOW);
    return ctx;
}

static LRESULT CALLBACK windowProc(HWND , UINT , WPARAM , LPARAM );

D2DContext *d2dl_initChildWindow(void *user, int width, int height,
                                  const char *window_title) {
    D2DContext *ctx = (D2DContext*)malloc(sizeof(D2DContext));
    if (ctx == NULL) {
        log_err_return_null("D2DContext allocation failed")
    }
    
    memset(ctx, 0, sizeof(D2DContext));
        
    if (globalOpenGUICount++ == 0) {
		WNDCLASS windowClass = {};
		windowClass.lpfnWndProc = windowProc;
		windowClass.lpszClassName = window_title;
		windowClass.hCursor = LoadCursor(NULL, IDC_ARROW);
		windowClass.style = CS_DBLCLKS;
		RegisterClass(&windowClass);
	}
	
    d2dl_createFactory(ctx);
    d2dl_createDirectWriteFactory(ctx);
    
    HWND hwnd = CreateWindow(window_title, window_title, WS_CHILDWINDOW | WS_CLIPSIBLINGS,
                             CW_USEDEFAULT, CW_USEDEFAULT, width,
                         height, GetDesktopWindow(), NULL, NULL, ctx);
    if (hwnd == NULL) {
        log_err_return_null("CreateWindow failed");
    }
    ctx->window = hwnd;
    ctx->user = user;
    d2dl_createRenderTarget(ctx);

    return ctx;
}

void d2dl_deinit(D2DContext *c) {
    DestroyWindow(c->window);
    c->solid_color->Release();
    c->rt->Release();
    c->factory->Release();
    c->dwrite_factory->Release();
    if (c->text_format)
        c->text_format->Release();
    free(c);
}

void d2dl_setParent(D2DContext *c, HWND parent) {
    SetParent(c->window, parent);
}

void d2dl_setVisible(D2DContext *c, bool visible) {
    ShowWindow(c->window, visible ? SW_SHOW : SW_HIDE);
}

static LRESULT CALLBACK windowProc(HWND window, UINT msg, WPARAM wParam, LPARAM lParam) {
    if (msg == WM_CREATE) {
        LPCREATESTRUCT pcs = (LPCREATESTRUCT)lParam;
        D2DContext *ctx = (D2DContext *)pcs->lpCreateParams;
        SetWindowLongPtr(window, GWLP_USERDATA, (LONG_PTR)ctx);
        return 1;
    }
    D2DContext *ctx = (D2DContext*)GetWindowLongPtr(window, GWLP_USERDATA);

    switch (msg) {
        case WM_DESTROY:
            PostQuitMessage(0);
            return 0;
        case WM_SIZE:
            if (!ctx->rt) {
                log_err_continue("Can't handle WM_SIZE yet");
                break;
            }
            d2dl_setRenderSize(ctx, {(float)LOWORD(lParam), (float)HIWORD(lParam)});
            return 0;
        case WM_PAINT:
            arbor_gui_render(ctx->user);
            return 0;
    }

    return DefWindowProc(window, msg, wParam, lParam);
}

File: src/gui/render/d2d.zig
const std = @import("std");
const windows = std.os.windows;
pub const HWND = windows.HWND;
const wchar = windows.WCHAR;

const toUtf16 = std.unicode.utf8ToUtf16LeAllocZ;
var g_unicode_buf: [128]u8 = .{0} ** 128;
var unicode_fba_impl = std.heap.FixedBufferAllocator.init(g_unicode_buf);
const unicode_fba = unicode_fba_impl.allocator();

// (big ol') TODO: Use marlersoft/zigwin32 as the Windows layer, eliminate
// our C++ library. Then these structs below will just be aliases of the direct
// Win32 types, and we dont' have to deal with compiling C and keeping 1,000
// different API declarations in sync.

pub const Context = opaque {
    pub fn deinit(self: *Context) void {
        d2dl_deinit(self);
    }

    pub fn setParent(self: *Context, parent: HWND) void {
        d2dl_setParent(self, parent);
    }

    pub fn setVisible(self: *Context, visible: bool) void {
        d2dl_setVisible(self, visible);
    }

    pub fn beginDrawing(self: *Context) void {
        d2dl_beginDrawing(self);
    }

    pub fn endDrawing(self: *Context) void {
        d2dl_endDrawing(self);
    }

    pub fn clear(self: *Context, color: Color) void {
        d2dl_clear(self, color);
    }

    pub fn drawRect(self: *Context, rect: Rect, stroke_width: f32, color: Color) void {
        d2dl_drawRect(self, rect, stroke_width, color);
    }
    pub fn fillRect(self: *Context, rect: Rect, color: Color) void {
        d2dl_fillRect(self, rect, color);
    }
    pub fn drawRoundedRect(self: *Context, rect: Rect, radius: f32, stroke_width: f32, color: Color) void {
        d2dl_drawRoundedRect(self, rect, radius, stroke_width, color);
    }
    pub fn fillRoundedRect(self: *Context, rect: Rect, radius: f32, color: Color) void {
        d2dl_fillRoundedRect(self, rect, radius, color);
    }
    pub fn drawEllipse(self: *Context, center: Point, rx: f32, ry: f32, stroke_width: f32, color: Color) void {
        d2dl_drawEllipse(self, center, rx, ry, stroke_width, color);
    }
    pub fn fillEllipse(self: *Context, center: Point, rx: f32, ry: f32, color: Color) void {
        d2dl_fillEllipse(self, center, rx, ry, color);
    }
    pub fn loadFont(self: *Context, font_name: []const u8, size: f32) void {
        const wtext: [:0]u16 = toUtf16(unicode_fba, font_name) catch |e| @panic(@errorName(e));
        d2dl_loadFont(self, wtext, size);
    }
    pub fn drawText(self: *Context, text: []const u8, rect: Rect, color: Color) void {
        const wtext: [:0]u16 = toUtf16(unicode_fba, text) catch |e| @panic(@errorName(e));
        d2dl_drawText(self, wtext, rect, color);
    }
    pub fn getRenderSize(self: *Context) Size {
        return d2dl_getRenderSize(self);
    }
    pub fn setRenderSize(self: *Context, size: Size) void {
        d2dl_setRenderSize(self, size);
    }
};

const WindowType = union(enum) {
    desktop,
    child: struct { user: *anyopaque },
};

pub fn init(
    width: i32,
    height: i32,
    window_title: [:0]const u8,
    window_type: WindowType,
) !*Context {
    switch (window_type) {
        .desktop => return d2dl_initDesktopWindow(width, height, window_title) orelse {
            return error.WindowInitFailed;
        },
        .child => |child| return d2dl_initChildWindow(
            child.user,
            width,
            height,
            window_title,
        ) orelse return error.WindowInitFailed,
    }
}

// Direct2D-compatible render data

pub const Color = extern struct {
    r: f32,
    g: f32,
    b: f32,
    a: f32 = 1,
};

pub const Point = extern struct {
    x: f32,
    y: f32,
};

pub const Size = extern struct {
    width: f32,
    height: f32,
};

pub const Rect = extern struct {
    left: f32,
    top: f32,
    right: f32,
    bottom: f32,
};

extern fn d2dl_initDesktopWindow(width: c_int, height: c_int, window_title: [*:0]const u8) ?*Context;
extern fn d2dl_initChildWindow(user: *anyopaque, width: c_int, height: c_int, window_title: [*:0]const u8) ?*Context;
extern fn d2dl_deinit(*Context) void;
extern fn d2dl_beginDrawing(*Context) void;
extern fn d2dl_endDrawing(*Context) void;
extern fn d2dl_clear(*Context, color: Color) void;
extern fn d2dl_drawRect(*Context, rect: Rect, stroke_width: f32, color: Color) void;
extern fn d2dl_fillRect(*Context, rect: Rect, color: Color) void;
extern fn d2dl_drawRoundedRect(*Context, rect: Rect, radius: f32, stroke_width: f32, color: Color) void;
extern fn d2dl_fillRoundedRect(*Context, rect: Rect, radius: f32, color: Color) void;
extern fn d2dl_drawEllipse(*Context, center: Point, rx: f32, ry: f32, stroke_width: f32, color: Color) void;
extern fn d2dl_fillEllipse(*Context, center: Point, rx: f32, ry: f32, color: Color) void;
extern fn d2dl_drawText(*Context, text: [*]const wchar, rect: Rect, color: Color) void;
extern fn d2dl_getRenderSize(*Context) Size;
extern fn d2dl_setRenderSize(*Context, size: Size) void;
extern fn d2dl_loadFont(*Context, font_name: [*]const wchar, size: f32) void;
extern fn d2dl_setParent(*Context, hwnd: HWND) void;
extern fn d2dl_setVisible(*Context, bool) void;
